sdk/typespec/typespec_client_core/src/http/pager.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use std::{future::Future, pin::Pin};
+
+use futures::{stream::unfold, Stream};
+use typespec::Error;
+
+use crate::http::Response;
+
+/// Represents a paginated result across multiple requests.
+#[pin_project::pin_project]
+pub struct Pager<T> {
+    #[pin]
+    #[cfg(not(target_arch = "wasm32"))]
+    stream: Pin<Box<dyn Stream<Item = Result<Response<T>, Error>> + Send>>,
+
+    #[pin]
+    #[cfg(target_arch = "wasm32")]
+    stream: Pin<Box<dyn Stream<Item = Result<Response<T>, Error>>>>,
+}
+
+impl<T> Pager<T> {
+    /// Creates a [`Pager<T>`] from a callback that will be called repeatedly to request each page.
+    ///
+    /// This method expect a callback that accepts a single `Option<C>` parameter, and returns a `(Response<T>, Option<C>)` tuple, asynchronously.
+    /// The `C` type parameter is the type of the continuation. It may be any [`Send`]able type.
```

Comment:
What was the point of getting rid of the `Continuable`? It was succinct, but was it creating more hassles? For an SDK, I'd rather we have clear names and intent, and unnamed (rather, unnamable) tuples erode that. We could use a concrete struct, but a trait like used before is flexible. Maybe unnecessarily so, though, which is why I'd love to understand more about why you replaced it with the tuple containing an `Option<C>`.

Reply:
The problem is: What type do you implement it on? The `Response<T>` has the headers and in Cosmos the continuation token is stored in a response header, so you'd want to do something like this, inside the Cosmos crate:

```rust
impl<T> Continuable for Response<QueryResults<T>> {
    type Continuation: String;

    fn continuation(&self) -> Option<Self::Continuation> {
        self.get_optional_string("x-ms-continuation")
    }
}
```

But that's illegal, because both `Continuable` and `Response<T>` are foreign types:

![image](https://github.com/user-attachments/assets/3f0baf42-7eab-4f30-845a-ea84964babdf)

You could do this:

```rust
impl<T> Continuable for QueryResults<T> {
    type Continuation: String;

    fn continuation(&self) -> Option<Self::Continuation> {
    }
}
```

But `QueryResults<T>` is the model type, and has no access to the headers.

Since the callback body can trivially extract the continuation, having it return the continuation value was the simplest option.

> It was succinct, but was it creating more hassles? For an SDK, I'd rather we have clear names and intent, and unnamed (rather, unnamable) tuples erode that.

I agree there. We have a few other options here. We could return a named struct of some kind instead of a tuple. Or we could use our own enum with named variants:

```rust
pub enum PaginationResult<T, C> { // Name TBD
    Continue { response: Response<T>, continuation: C },
    Finish { response: Response<T> },
}
```

I'm not totally sure those are much better, since they make the callback more verbose, but I'm game.

Reply:
Another note: Most of the time, your continuation token is _already an `Option<C>`_ (for whatever `C` you're using as a continuation value). For example, in Cosmos the code currently looks like this:

```rust
Pager::from_callback(move |continuation| {
    let pipeline = pipeline.clone();
    let mut req = base_req.clone();
    async move {
        if let Some(continuation) = continuation {
            req.insert_header(constants::CONTINUATION, continuation);
        }

        let resp = pipeline
            .send(Context::new(), &mut req, ResourceType::Items)
            .await?;
        let continuation_token =
            resp.headers().get_optional_string(&constants::CONTINUATION);

        Ok((resp, continuation_token))
    }
})
```

I don't know that trying to avoid the tuple gets us something clearer, but I'm still open to it if you disagree.

Reply:
I like something like `PagerResult<T, C>` where it's coercible to a `Result<T>` and you can get an `Option<C>` from the `Continue` variant. Seems `None` would be equally fine for either variant. That wouldn't look to bad, would it?

Thanks for the context about the reason for change. Makes sense.

Reply:
@JeffreyRichter do you think accepting and returning a continuation token will work in all paging cases? Seems to me it would, but you might have some insights on other paging workflows where that may not work.

In this design, we can get a continuation token from headers or body. But what about those services that still use `top` and `skip`? Do we manifest a continuation token that we use to keep track of the sliding window of results e.g., doing something akin got `skip += top`?

Reply:
It's important to note that because the return type is `Option<C>` (instead of something concrete like `Option<String>`), `top/skip` continuation would absolutely work here. The callback could extract the number that need to be skipped for the next call and return that as their continuation value.

I can look at something using a concrete type. I'm a little concerned we'll be adding unnecessary verbosity to APIs that are used _predominantly_ by service client authors (as they're the main audience who are _creating_ `Pager<T>` instances) but I can spend some more cycles on it.